README.md
# Campaign Compendium

CHANGELOG.md
## Next Release - TBC

### Added

### Changed

### Fixed

## [1.2.1] - 24/07/2022

### Added

### Changed

### Fixed

- Fixed issue with "Add Quest" button not appearing when no quests exist on the campaign
- Fixed issue with session update to render the line breaks entered and saved previously

## [1.2.0] - 23/07/2022
### Added

- Added new `POST` `/campaigns/:id/quests` endpoint for users to create quests on a campaign
- Added new `GET` `/campaigns/:id/quests` endpoint for users to fetch quests on a campaign
- Added new `DELETE` `/quests/:id` endpoint for users to delete a quest and all associated milestones
- Added new `PATCH` `/quests/:id` endpoint for users to edit a quest
- Added new `POST` `/quest-milestones/:id/complete` endpoint for users to mark a quest as complete
- Added new `POST` `/quest-milestones/:id/incomplete` endpoint for users to mark a quest as incomplete
- Added navigation to `quests` feature from campaign page
- Added UI elements to create, edit, and delete quests, and tick off objectives as they are completed
- Swapped out rich text formatting for an ordinary text input on session updates due to lack of mobile support

### Changed

### Fixed

- Fixed a bug with edit campaign description which populated the edit box with the campaign name.

## [1.1.2] - 2022-06-09
### Added

### Changed

### Fixed

- Fixed issue with delete note modal not closing on confirmation
- Fixed issue with removing a player from a campaign
- Ensure session update modal state is reset when adding a new session update

## [1.1.1] - 2022-06-05
### Added

### Changed

### Fixed

- Fixed issue with list of characters on notes filter persisting from a previously-viewed campaign

## [1.1.0] - 2022-06-05
### Added

- Added new `GET` `/campaigns/:id/characters` endpoint so all characters (including deleted ones) on a campaign can be fetched
- Added new `POST` `/campaigns/:id/notes` endpoint for users to create notes on campaigns
- Added new `GET` `/campaigns/:id/notes` endpoint for DMs to fetch all notes on a campaign
- Added new `GET` `/campaigns/:id/notes/created-by-me` endpoint for players to fetch their notes on a campaign
- Added new `PATCH` `/notes/:id` endpoint for users to update the notes they've created
- Added new `DELETE` `/notes/:id` endpoint for users to delete their notes
- Added new `POST` `/campaigns/:id/session-updates` endpoint for DMs to create new session updates
- Added new `PATCH` `/session-updates/:id` endpoint for DMs to update their session updates
- Added new `GET` `/campaigns/:id/session-updates` endpoint for users to get all session updates on a campaign
- Added new `DELETE` `/session-updates/:id` endpoint so a DM can delete a session update
- Added new menu to UI to allow users to navigate between campaign overview, session updates, and notes
- Added new campaign page for users to interact with session updates
- Added new campaign page for users to interact with notes

### Changed

- Changed character card to display a tombstone next to the character name if the character is dead

### Fixed


## [1.0.0] - 2022-04-15
### Added

- Added `SVG` icons for character cards
- Added an image picker to allow users to set and change their character image
- Added colour picker to allow users to set and change their character colour
- Added privacy policy and links to privacy policy from login page and user menu

### Changed

### Fixed

## [0.0.3] - 2022-04-14
### Added

- Added `deletedAt` property to `characters` so they can be soft-deleted
- Added new `POST` `/campaigns/:id/user/:id` endpoint which will soft-deleted all characters associated with that user and campaign
- Added DM option on player kebab menu to allow a DM to delete a single character
- Added DM option on player kebab menu to allow a DM to remove a player (deletes all characters belonging to that player in that campaign)
- Added DM option on player kebab menu to re-invite a player to join the campaign again with a new character
- Added prompt at the top of the player campaign view that displays when a DM has reinvited them to join the campaign
- Added player graveyard to display dead characters

### Changed

- Changed `DELETE` `/characters/:id` to perform an update on a character - setting the `deletedAt` property instead of deleting the document
- Changed `DELETE` `/characters/:id` to return a `404` if character for given ID exists but has a populated `deletedAt` property
- Changed `POST` `/characters/:id/kill` to return a `404` if character for given ID exists but has a populated `deletedAt` property
- Changed `POST` `/characters/:id/revive` to return a `404` if character for given ID exists but has a populated `deletedAt` property
- Changed `POST` `/characters/:id/revive` to return a `400` if player already has an `active` or `invited` character in this campaign already
- Changed `POST` `/campaigns/:id/invite` to return a `400` if a character with the same user ID already exists on the campaign at status `active` or `invited`
- Changed `GET` `/campaigns/:id` to not return characters with a populated `deletedAt` property
- Changed `GET` `/campaigns/:id/player` to not return characters with a populated `deletedAt` property
- Only display a campaign in a player's `invited` list if they only have a single character in that campaign at status `invited`
- Display all other campaigns a player is in, in the `active` list if they have any other configuration of characters (1 `active` 0-* `dead` characters or 1 `invited` and 1-* `dead` characters )

### Fixed

## [0.0.2] - 2022-04-10
### Added
- Add helper text to campaign `name` and `description` fields to show required error message when creating a campaign
- Added ability to invite a player to join an existing campaign

### Changed

- Restrict `PATCH` `/campaigns/:id` to only the `creator` of that campaign
- Restrict `POST` `/campaigns/:id/invite` to only the `creator` of that campaign

src/routers/note.js
const express = require('express');
const router = new express.Router();
const mongoose = require('mongoose');
const Note = require('../models/note');
const Character = require('../models/character');
const Campaign = require('../models/campaign');

// POST
// POST endpoint to create a new note - /campaigns/:id/notes
// only the creator of a campaign can create a note, or a user who has a non-deleted character on the given campaign
router.post("/campaigns/:id/notes", async (req, res) => {
    if (req.user) {
        try {
            const _id = req.params.id;
            if (!mongoose.Types.ObjectId.isValid(_id)) return res.status(404).send('A campaign could not be found for the given ID');

            const campaign = await Campaign.findById(_id);
            if (!campaign) return res.status(404).send('A campaign could not be found for the given ID');

            const character = await Character.find({ 'campaignId': _id, 'userId': req.user._id, 'deletedAt': '' });
            // check if the user sending the request is the creator of this campaign OR has a non-deleted character on the given campaign
            if (!campaign.createdBy.equals(req.user._id) && (!character || character.length === 0)) return res.status(401).send('You are not authorised to add notes to this campaign');

            const { content } = req.body;

            if (!content) return res.status(400).send('Note content is required');

            // get the related character from the request
            let newRelatedCharacter = req.body.hasOwnProperty('relatedCharacter') ? req.body.relatedCharacter : undefined;

            // if no related character param exists and the message has not come from the requestor, return an error
            if (!newRelatedCharacter && !campaign.createdBy.equals(req.user._id)) return res.status(400).send('You must specify a related character');

            // since no error was returned, if there's no related character this must be the DM so set the value to "DM"
            if (!newRelatedCharacter) newRelatedCharacter = "DM";

            // if the related character is not the DM, check that this character exists on the campaign
            if (newRelatedCharacter !== "DM") {
                const campaignCharacter = await Character.find({ 'campaignId': _id, 'userId': req.user._id, 'deletedAt': '', '_id': newRelatedCharacter });
                if (!campaignCharacter || campaignCharacter.length === 0) return res.status(404).send('You do not have a character on this campaign for the given related character ID');
                if (campaignCharacter.status === 'invited') return res.status(400).send('Notes cannot be related to an invited character');
            }

            const note = new Note({
                content,
                relatedCharacter: newRelatedCharacter,
                createdAt: new Date().toISOString(),
                createdBy: req.user._id,
                campaignId: _id
            });

            await note.save();
            res.status(201).send(note);
        } catch (e) {
            res.status(404).send(e);
        }
    } else {
        res.status(401).send({});
    }
});


// GET - all notes
// GET endpoint to fetch all notes on a campaign - /campaigns/:id/notes
// only the DM is authorised to this endpoint
// paginated
// param options so DM can specify the character(s) to filter on (including themselves with the string "DM")
// also date range from and to params to filter between a date range
router.get("/campaigns/:id/notes", async (req, res) => {
    if (req.user) {
        // get all session updates for the campaign
        try {
            const _id = req.params.id;
            if (!mongoose.Types.ObjectId.isValid(_id)) return res.status(404).send('A campaign could not be found for the given ID');

            const campaign = await Campaign.findById(_id);
            if (!campaign) return res.status(404).send('A campaign could not be found for the given ID');

            if (!campaign.createdBy.equals(req.user._id)) {
                return res.status(401).send('You are not authorised to get all notes on this campaign');
            }

            // access query params
            const relatedCharactersParam = req.query.relatedCharacter; // will be a string or array of strings of character IDs potentially with "DM" in the mix | undefined
            const fromParam = req.query.from; // will be string of date ISO string | undefined
            const toParam = req.query.to; // will be string of date ISO string | undefined
            const page = parseInt(req.query.page, 10);
            const pageSize = parseInt(req.query.pageSize, 10);

            if (!page) return res.status(400).send('Param page is required');
            if (!pageSize) return res.status(400).send('Param pageSize is required');
            if (!fromParam) return res.status(400).send('Param from is required');
            if (!toParam) return res.status(400).send('Param to is required');

            // if the param is already an array, use it, otherwise create an array
            let filterCharactersArray = [];
            if (relatedCharactersParam && Array.isArray(relatedCharactersParam)) filterCharactersArray = relatedCharactersParam;
            else if (relatedCharactersParam) filterCharactersArray = [relatedCharactersParam];

            // base conditions for the find query
            const conditions = [
                { campaignId: _id },
                { createdAt: { $gt: fromParam } },
                { createdAt: { $lt: toParam }}
            ];

            // if filterCharactersArray is empty then don't filter on characters, just get all notes in the date range
            if (filterCharactersArray.length > 0) conditions.push({ relatedCharacter: { $in: filterCharactersArray } });

            const notes = await Note.find({ $and: conditions })
                .sort({ createdAt: 'desc' })
                .skip((page - 1) * pageSize)
                .limit(pageSize)
                .lean();

            const notesCount = await Note.find({ $and: conditions }).count();

            const notesWithCharacters = [];

            if (notes) {
                // we have notes so we need to get the character associated with the note
                for (const note of notes) {
                    const newNote = note;
                    if (note.relatedCharacter !== "DM") {
                        const noteCharacter = await Character.findById(note.relatedCharacter);
                        if (noteCharacter) newNote.character = noteCharacter;
                    }
                    notesWithCharacters.push(newNote);
                }
            }

            res.status(200).send({ data: notesWithCharacters, count: notesCount });
        } catch (e) {
            res.status(404).send(e);
        }
    } else {
        res.status(401).send({});
    }
});


// GET - player notes
// GET endpoint to fetch only "my" notes on a campaign - /campaigns/:id/notes/created-by-me
// this just returns any notes created by the logged in user on the given campaign
// the logged in user must have a non-deleted character on the given campaign
// this is paginated
router.get("/campaigns/:id/notes/created-by-me", async (req, res) => {
    if (req.user) {
        // get all session updates for the campaign
        try {
            const _id = req.params.id;
            if (!mongoose.Types.ObjectId.isValid(_id)) return res.status(404).send('A campaign could not be found for the given ID');

            const campaign = await Campaign.findById(_id);
            if (!campaign) return res.status(404).send('A campaign could not be found for the given ID');

            const character = await Character.find({ 'campaignId': _id, 'userId': req.user._id, 'deletedAt': '' });
            if (!character || character.length === 0) return res.status(400).send('You do not have a character on this campaign');

            // access query params
            const page = parseInt(req.query.page, 10);
            const pageSize = parseInt(req.query.pageSize, 10);

            if (!page) return res.status(400).send('Param page is required');
            if (!pageSize) return res.status(400).send('Param pageSize is required');

            const notes = await Note.find({ 'campaignId': _id, 'createdBy': req.user._id })
                .sort({ createdAt: 'desc' })
                .skip((page - 1) * pageSize)
                .limit(pageSize)
                .lean();

            const notesCount = await Note.find({ 'campaignId': _id, 'createdBy': req.user._id }).count();

            const notesWithCharacters = [];

            if (notes) {
                // we have notes so we need to get the character associated with the note
                for (const note of notes) {
                    const newNote = note;
                    if (note.relatedCharacter !== "DM") {
                        const noteCharacter = await Character.findById(note.relatedCharacter);
                        if (noteCharacter) newNote.character = noteCharacter;
                    }
                    notesWithCharacters.push(newNote);
                }
            }

            res.status(200).send({ data: notesWithCharacters, count: notesCount });
        } catch (e) {
            res.status(404).send(e);
        }
    } else {
        res.status(401).send({});
    }
});

// PATCH
// PATCH endpoint so that the creator of a note can edit it - /notes/:id
// only the creator of the note can edit it
// the related character cannot be changed to DM or away from DM
router.patch("/notes/:id", async (req, res) => {
    if (req.user) {
        const _id = req.params.id;
        try {
            if (!mongoose.Types.ObjectId.isValid(_id)) return res.status(404).send('A note could not be found for the given ID');

            const note = await Note.findById(_id);

            if (!note) return res.status(404).send('A note could not be found for the given id');

            if (!note.createdBy.equals(req.user._id)) {
                return res.status(401).send('You are not authorised to edit this note');
            }

            const newRelatedCharacter = req.body.hasOwnProperty('relatedCharacter') ? req.body.relatedCharacter : undefined;
            if (newRelatedCharacter === "DM" && note.relatedCharacter !== "DM") return res.status(400).send('You cannot move this note to being related to the DM');
            if (newRelatedCharacter !== "DM" && note.relatedCharacter === "DM") return res.status(400).send('You cannot move this note from being related to the DM');

            if (newRelatedCharacter) note.relatedCharacter = req.body.relatedCharacter;
            if (req.body.hasOwnProperty('content')) note.content = req.body.content;
            await note.save();
            res.send(note);
        } catch (e) {
            res.status(400).send(e);
        }
    } else {
        res.status(401).send({});
    }
});

// DELETE
// DELETE endpoint to delete a note - /notes/:id
// only the creator of the note can edit it
router.delete("/notes/:id", async (req, res) => {
    if (req.user) {
        const _id = req.params.id;
        try {
            if (!mongoose.Types.ObjectId.isValid(_id)) return res.status(404).send('A note could not be found for the given ID');

            const note = await Note.findById(_id);

            if (!note) return res.status(404).send('A note could not be found for the given id')

            if (!note.createdBy.equals(req.user._id)) {
                return res.status(401).send('You are not authorised to delete this note');
            }

            const deletedNote = await Note.deleteOne({ _id });
            res.send(deletedNote);
        } catch (e) {
            res.status(400).send(e);
        }
    } else {
        res.status(401).send({});
    }
});

module.exports = router;
